File: android/measure/src/test/java/sh/measure/android/tracing/MsrSpanTest.kt
package sh.measure.android.tracing

import org.junit.Assert
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.times
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.verify
import sh.measure.android.attributes.AttributesBuilder
import sh.measure.android.fakes.FakeSessionManager
import sh.measure.android.fakes.FakeTraceSampler
import sh.measure.android.fakes.NoopLogger
import sh.measure.android.utils.AndroidTimeProvider
import sh.measure.android.utils.IdProviderImpl
import sh.measure.android.utils.RandomizerImpl
import sh.measure.android.utils.TestClock
import java.time.Duration

class MsrSpanTest {
    private val logger = NoopLogger()
    private val testClock = TestClock.create()
    private val timeProvider = AndroidTimeProvider(testClock)
    private val idProvider = IdProviderImpl(randomizer = RandomizerImpl())
    private val spanProcessor = mock<SpanProcessor>()
    private val sessionManager = FakeSessionManager()
    private val traceSampler = FakeTraceSampler()

    @Test
    fun `startSpan sets parent span if provided`() {
        val parentSpan = MsrSpan(
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            isSampled = true,
            name = "parent-span",
            spanId = "span-id",
            traceId = "trace-id",
            parentId = null,
            sessionId = sessionManager.getSessionId(),
            startTime = 1000,
        )
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = parentSpan,
            timestamp = null,
        )

        Assert.assertEquals(parentSpan.spanId, span.parentId)
    }

    @Test
    fun `startSpan sets current timestamp`() {
        val epochTime = testClock.epochTime()
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
            timestamp = null,
        ) as MsrSpan
        Assert.assertEquals(epochTime, span.startTime)
    }

    @Test
    fun `startSpan sets timestamp if provided`() {
        val timestamp = 10000L
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
            timestamp = timestamp,
        ) as MsrSpan
        Assert.assertEquals(timestamp, span.startTime)
    }

    @Test
    fun `startSpan triggers span processor onStart`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ) as MsrSpan

        verify(spanProcessor, times(1)).onStart(span)
    }

    @Test
    fun `default span status is unset`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ) as MsrSpan
        Assert.assertEquals(SpanStatus.Unset, span.getStatus())
    }

    @Test
    fun `setStatus updates the span status`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ).setStatus(SpanStatus.Ok) as MsrSpan
        Assert.assertEquals(SpanStatus.Ok, span.getStatus())
    }

    @Test
    fun `setName updates the span name`() {
        val span = MsrSpan.startSpan(
            "initial-span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ).setName("updated-span-name") as MsrSpan
        Assert.assertEquals("updated-span-name", span.name)
    }

    @Test
    fun `hasEnded for active span`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        Assert.assertEquals(false, span.hasEnded())
    }

    @Test
    fun `hasEnded for ended span`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ).end()
        Assert.assertEquals(true, span.hasEnded())
    }

    @Test
    fun `end updates the span duration`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        testClock.advance(Duration.ofMillis(1000))
        val duration = span.end().getDuration()

        Assert.assertEquals(1000, duration)
    }

    @Test
    fun `end triggers span processor onEnding and onEnded`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ).end() as MsrSpan

        spanProcessor.inOrder {
            verify().onEnding(span)
            verify().onEnded(span)
        }
    }

    @Test
    fun `setCheckpoint adds checkpoint to span`() {
        val expectedCheckpoint = Checkpoint("name", timeProvider.now())
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ) as MsrSpan
        span.setCheckpoint(expectedCheckpoint.name)

        Assert.assertEquals(1, span.checkpoints.size)
        Assert.assertEquals(expectedCheckpoint, span.checkpoints.first())
    }

    @Test
    fun `setEvent on ended span is a no-op`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ).end() as MsrSpan
        span.setCheckpoint("event-id")

        Assert.assertEquals(0, span.checkpoints.size)
    }

    @Test
    fun `setAttribute adds attribute to span`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ) as MsrSpan
        span.setAttribute("key", "value")

        Assert.assertEquals("value", span.getUserDefinedAttrs()["key"])
    }

    @Test
    fun `setAttribute on ended span is a no-op`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ).end() as MsrSpan
        span.setAttribute("key", "value")

        Assert.assertEquals(0, span.attributes.size)
    }

    @Test
    fun `setAttributes adds multiple attributes to span`() {
        val attributes = AttributesBuilder().put("key1", "value1").put("key2", "value2").build()
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ) as MsrSpan
        span.setAttributes(attributes)

        Assert.assertEquals("value1", span.getUserDefinedAttrs()["key1"])
        Assert.assertEquals("value2", span.getUserDefinedAttrs()["key2"])
    }

    @Test
    fun `removeAttribute removes attribute from span`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        ) as MsrSpan
        span.setAttribute("key", "value")
        span.removeAttribute("key")

        Assert.assertEquals(0, span.getUserDefinedAttrs().size)
    }

    @Test
    fun `duration is 0 for active span`() {
        val span = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        testClock.advance(Duration.ofMillis(1000))
        val duration = span.getDuration()

        Assert.assertEquals(0, duration)
    }

    @Test
    fun `startSpan sets sampling state for root span based on trace sampler`() {
        traceSampler.setSampled(true)
        val span1 = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        Assert.assertTrue(span1.isSampled)

        traceSampler.setSampled(false)
        val span2 = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        Assert.assertFalse(span2.isSampled)
    }

    @Test
    fun `startSpan samples child span if parent span is sampled`() {
        traceSampler.setSampled(true)
        val sampledParentSpan = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        traceSampler.setSampled(false)
        val childSpan = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = sampledParentSpan,
        )

        Assert.assertTrue(childSpan.isSampled)
    }

    @Test
    fun `startSpan does not sample child span if parent span is not sampled`() {
        traceSampler.setSampled(false)
        val unsampledParentSpan = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        traceSampler.setSampled(true)
        val childSpan = MsrSpan.startSpan(
            "span-name",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = unsampledParentSpan,
        )

        Assert.assertFalse(childSpan.isSampled)
    }

    @Test
    fun `startSpan initializes parent ID and trace ID correctly`() {
        val parentSpan = MsrSpan.startSpan(
            "parent-span",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        val childSpan = MsrSpan.startSpan(
            "child-span",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = parentSpan,
        )

        Assert.assertEquals(parentSpan.traceId, childSpan.traceId)
        Assert.assertEquals(parentSpan.spanId, childSpan.parentId)
    }

    @Test
    fun `setParent updates parent ID and trace ID correctly`() {
        val parentSpan = MsrSpan.startSpan(
            "parent-span",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )
        val childSpan = MsrSpan.startSpan(
            "child-span",
            logger = logger,
            timeProvider = timeProvider,
            spanProcessor = spanProcessor,
            sessionManager = sessionManager,
            traceSampler = traceSampler,
            idProvider = idProvider,
            parentSpan = null,
        )

        childSpan.setParent(parentSpan)

        Assert.assertEquals(parentSpan.traceId, childSpan.traceId)
        Assert.assertEquals(parentSpan.spanId, childSpan.parentId)
    }
}

File: flutter/packages/measure_flutter/test/tracing/msr_span_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:measure_flutter/measure_flutter.dart';
import 'package:measure_flutter/src/time/time_provider.dart';
import 'package:measure_flutter/src/tracing/msr_span.dart';

import '../utils/fake_id_provider.dart';
import '../utils/fake_span_processor.dart';
import '../utils/fake_trace_sampler.dart';
import '../utils/noop_logger.dart';
import '../utils/test_clock.dart';

void main() {
  group('MsrSpan', () {
    late NoopLogger logger;
    late TestClock testClock;
    late FlutterTimeProvider timeProvider;
    late FakeSpanProcessor spanProcessor;
    late FakeTraceSampler traceSampler;
    late FakeIdProvider idProvider;

    setUp(() {
      logger = NoopLogger();
      testClock = TestClock.create();
      timeProvider = FlutterTimeProvider(testClock);
      spanProcessor = FakeSpanProcessor();
      traceSampler = FakeTraceSampler();
      idProvider = FakeIdProvider();
    });

    group('startSpan', () {
      test('sets parent span if provided', () {
        final parentSpan = MsrSpan(
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          isSampled: true,
          name: 'test',
          spanId: 'spanId',
          traceId: 'traceId',
          parentId: 'parentId',
          startTime: 123,
        );

        final span = MsrSpan.startSpan(
          name: 'test',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          parentSpan: parentSpan,
          idProvider: idProvider,
          traceSampler: traceSampler,
        );

        expect(span.parentId, parentSpan.spanId);
      });

      test('sets current timestamp', () {
        final epochTime = testClock.epochTime();
        final span = MsrSpan.startSpan(
          name: 'test',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
          timestamp: null,
        ) as MsrSpan;

        expect(span.startTime, epochTime);
      });

      test('sets timestamp if provided', () {
        final timestamp = 10000;
        final span = MsrSpan.startSpan(
          name: 'test',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
          timestamp: timestamp,
        ) as MsrSpan;

        expect(span.startTime, timestamp);
      });

      test('triggers span processor onStart', () {
        MsrSpan.startSpan(
          name: 'test',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
        );

        expect(spanProcessor.startedSpanCount, 1);
      });

      test('sets sampling state for root span based on trace sampler', () {
        traceSampler.overrideShouldSample = true;
        final span1 = MsrSpan.startSpan(
          name: 'span-name',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        expect(span1.isSampled, true);

        traceSampler.overrideShouldSample = false;
        final span2 = MsrSpan.startSpan(
          name: 'span-name',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        expect(span2.isSampled, false);
      });

      test('samples child span if parent span is sampled', () {
        traceSampler.overrideShouldSample = true;
        final sampledParentSpan = MsrSpan.startSpan(
          name: 'span-name',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        traceSampler.overrideShouldSample = false;
        final childSpan = MsrSpan.startSpan(
          name: 'span-name',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: sampledParentSpan,
        ) as MsrSpan;

        expect(childSpan.isSampled, true);
      });

      test('does not sample child span if parent span is not sampled', () {
        traceSampler.overrideShouldSample = false;
        final unsampledParentSpan = MsrSpan.startSpan(
          name: 'span-name',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        traceSampler.overrideShouldSample = true;
        final childSpan = MsrSpan.startSpan(
          name: 'span-name',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: unsampledParentSpan,
        ) as MsrSpan;

        expect(childSpan.isSampled, false);
      });

      test('initializes parent ID and trace ID correctly', () {
        final parentSpan = MsrSpan.startSpan(
          name: 'parent-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        final childSpan = MsrSpan.startSpan(
          name: 'child-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: parentSpan,
        ) as MsrSpan;

        expect(childSpan.parentId, parentSpan.spanId);
        expect(childSpan.traceId, parentSpan.traceId);
      });
    });

    group('span status', () {
      test('default span status is unset', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        expect(span.getStatus(), SpanStatus.unset);
      });

      test('setStatus updates the span status', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.setStatus(SpanStatus.ok);
        expect(span.getStatus(), SpanStatus.ok);

        span.setStatus(SpanStatus.error);
        expect(span.getStatus(), SpanStatus.error);
      });
    });

    group('span name', () {
      test('setName updates the span name', () {
        final span = MsrSpan.startSpan(
          name: 'original-name',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        expect(span.name, 'original-name');

        span.setName('updated-name');
        expect(span.name, 'updated-name');
      });
    });

    group('span lifecycle', () {
      test('hasEnded for active span', () {
        final span = MsrSpan.startSpan(
          name: 'active-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        expect(span.hasEnded(), false);
      });

      test('hasEnded for ended span', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.end();
        expect(span.hasEnded(), true);
      });
    });

    group('end', () {
      test('updates the span duration', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        // Advance time and end the span
        testClock.advance(Duration(milliseconds: 500));
        span.end();

        expect(span.getDuration(), 500);
      });

      test('triggers span processor onEnding and onEnded', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.end();

        expect(spanProcessor.endingSpanCount, 1);
        expect(spanProcessor.endedSpanCount, 1);
      });

      test('uses the timestamp provided', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        // End the span with explicit timestamp
        span.end(timestamp: testClock.epochTime() + 10000);

        expect(span.getDuration(), 10000);
      });
    });

    group('checkpoints', () {
      test('setCheckpoint adds checkpoint to span', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.setCheckpoint('checkpoint-1');
        span.setCheckpoint('checkpoint-2');

        expect(span.checkpoints.length, 2);
        expect(span.checkpoints[0].name, 'checkpoint-1');
        expect(span.checkpoints[1].name, 'checkpoint-2');
      });

      test('setEvent on ended span is a no-op', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.end();
        final checkpointsBeforeEvent = span.checkpoints.length;

        span.setCheckpoint('event-after-end');

        expect(span.checkpoints.length, checkpointsBeforeEvent);
      });
    });

    group('attributes', () {
      test('setAttribute adds attribute to span', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.setAttributeString('key1', 'value1');
        span.setAttributeInt('key2', 42);

        expect(span.getUserDefinedAttrs()['key1'], 'value1');
        expect(span.getUserDefinedAttrs()['key2'], 42);
      });

      test('setAttribute on ended span is a no-op', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.setAttributeString('before-end', 'value');
        span.end();

        final attributesBeforeSet = Map.from(span.attributes);
        span.setAttributeString('after-end', 'value');

        expect(span.attributes, attributesBeforeSet);
        expect(span.attributes.containsKey('after-end'), false);
      });

      test('setAttributes adds multiple attributes to span', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        final attributes = {
          'key1': StringAttr('value1'),
          'key2': StringAttr('value2'),
          'key3': IntAttr(123),
        };

        span.setAttributes(attributes);

        expect(span.getUserDefinedAttrs()['key1'], 'value1');
        expect(span.getUserDefinedAttrs()['key2'], 'value2');
        expect(span.getUserDefinedAttrs()['key3'], 123);
      });

      test('removeAttribute removes attribute from span', () {
        final span = MsrSpan.startSpan(
          name: 'test-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        span.setAttributeString('key1', 'value1');
        span.setAttributeString('key2', 'value2');

        expect(span.getUserDefinedAttrs().containsKey('key1'), true);

        span.removeAttribute('key1');

        expect(span.getUserDefinedAttrs().containsKey('key1'), false);
        expect(span.getUserDefinedAttrs().containsKey('key2'), true);
      });
    });

    group('duration', () {
      test('is 0 for active span', () {
        final span = MsrSpan.startSpan(
          name: 'active-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        expect(span.getDuration(), 0);
      });
    });

    group('parent relationship', () {
      test('setParent updates parent ID and trace ID correctly', () {
        final parentSpan = MsrSpan.startSpan(
          name: 'parent-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        final childSpan = MsrSpan.startSpan(
          name: 'child-span',
          logger: logger,
          spanProcessor: spanProcessor,
          timeProvider: timeProvider,
          idProvider: idProvider,
          traceSampler: traceSampler,
          parentSpan: null,
        ) as MsrSpan;

        childSpan.setParent(parentSpan);

        expect(childSpan.parentId, parentSpan.spanId);
        expect(childSpan.traceId, parentSpan.traceId);
      });
    });
  });
}
